### PR TITLE
Implemented heatmap

### PR DESCRIPTION
**Pull request by DeyuChen**
Implemented a basic heatmap using the new refactored structure. Right now it is using exactly the same queries (not aggregated) for the pinmap.

### CODE REVIEW

**Comment by DeyuChen** [comment]
screenshot from 2018-03-09 16-45-32

File: examples/twittermap/web/conf/application.conf [hunk 1]
```diff
@@ -115,6 +115,12 @@ us.city.path = "/public/data/city.sample.json"
 
 tweet.url = "http://twitterMapIndex-search-proxy.herokuapp.com/search/tweets?q=%s"
 
+# Heatmap parameters
+# How many recent days of the time range predicate of each heatmap query.
+# heatmap.samplingDayRange = 30
```
Comment (waans11) [inline examples/twittermap/web/conf/application.conf]: Let's also mention that 30 is the default value.

File: examples/twittermap/web/conf/application.conf [hunk 2]
```diff
@@ -115,6 +115,12 @@ us.city.path = "/public/data/city.sample.json"
...
+# How many tweets at most returned for each heatmap query.
+# heatmap.samplingLimit = 5000
```
Comment (waans11) [inline examples/twittermap/web/conf/application.conf]: Let's also mention that 5000 is the default value.

File: examples/twittermap/web/public/javascripts/common/services.js [hunk 3]
```diff
@@ -368,6 +370,62 @@ angular.module('cloudberry.common', ['cloudberry.mapresultcache'])
...
+              },
+              transform: {
+                wrap: {
+                  id: "heat",
+                  category: "heat"
+                }
+              }
+            }));
+
+            // for the time histogram
```
Comment (waans11) [inline examples/twittermap/web/public/javascripts/common/services.js]: Why can't we use the same query from CountMap or PinMap?

File: examples/twittermap/web/public/javascripts/common/services.js [hunk 4]
```diff
@@ -471,6 +529,17 @@ angular.module('cloudberry.common', ['cloudberry.mapresultcache'])
             break;
           case "batchPointMapRequest":
             break;
+          case "heat":
```
Comment (waans11) [inline examples/twittermap/web/public/javascripts/common/services.js]: case "heatMapResult"

File: examples/twittermap/web/public/javascripts/common/services.js [hunk 5]
```diff
@@ -471,6 +529,17 @@ angular.module('cloudberry.common', ['cloudberry.mapresultcache'])
...
+            if(angular.isArray(result.value)) {
+              cloudberryService.commonTweetResult = result.value[0].slice(0, defaultSamplingSize - 1);
+              cloudberryService.heatmapMapResult = result.value[0];
+            }
+            break;
+          case "heatTime":
```
Comment (waans11) [inline examples/twittermap/web/public/javascripts/common/services.js]: Why this heatTime is necessary...

File: examples/twittermap/web/public/javascripts/map/controllers.js [hunk 6]
```diff
@@ -425,6 +425,24 @@ angular.module('cloudberry.map', ['leaflet-directive', 'cloudberry.common','clou
         }
       });
     }
+    
+    //For randomize coordinates by bounding_box
+    var gseed;
```
Comment (waans11) [inline examples/twittermap/web/public/javascripts/map/controllers.js]: Can we set a different name for this variable? (more meaningful)

File: examples/twittermap/web/public/javascripts/map/controllers.js [hunk 7]
```diff
@@ -425,6 +425,24 @@ angular.module('cloudberry.map', ['leaflet-directive', 'cloudberry.common','clou
...
+
+    function CustomRandom() {
```
Comment (waans11) [inline examples/twittermap/web/public/javascripts/map/controllers.js]: Let's put a simple comment for each function.

File: examples/twittermap/web/public/javascripts/map/heatmap.js [hunk 8]
```diff
@@ -1,28 +1,119 @@
...
+          color: '#666',
+          fillOpacity: 0
+        },
+        colors: [ '#ffffff', '#92d1e1', '#4393c3', '#2166ac', '#f4a582', '#d6604d', '#b2182b'],
+        sentimentColors: ['#ff0000', '#C0C0C0', '#00ff00']
+      });
     }
     
     function cleanHeatMap() {
+      if ($scope.heatLayer){
```
Comment (waans11) [inline examples/twittermap/web/public/javascripts/map/heatmap.js]: heatMapLayer

File: examples/twittermap/web/public/javascripts/map/heatmap.js [hunk 9]
```diff
@@ -1,28 +1,119 @@
...
+      $scope.resetZoomFunction(onEachFeature);
+      $scope.resetDragFunction(onEachFeature);
+
+      if (!$scope.heat){
+        $scope.heat = L.heatLayer([]);
+      }
     }
     
     function drawHeatMap(result) {
+      function setHeatPoints(points) {
```
Comment (waans11) [inline examples/twittermap/web/public/javascripts/map/heatmap.js]: setHeatMapPoints

File: examples/twittermap/web/conf/application.conf (hunk 1 above)
Comment (DeyuChen) [reply examples/twittermap/web/conf/application.conf]: Done

File: examples/twittermap/web/public/javascripts/common/services.js (hunk 3 above)
Comment (DeyuChen) [reply examples/twittermap/web/public/javascripts/common/services.js]: Agree. It is more like a place holder for now.

File: examples/twittermap/web/public/javascripts/common/services.js (hunk 4 above)
Comment (DeyuChen) [reply examples/twittermap/web/public/javascripts/common/services.js]: Done

File: examples/twittermap/web/public/javascripts/common/services.js (hunk 5 above)
Comment (DeyuChen) [reply examples/twittermap/web/public/javascripts/common/services.js]: It is there for the same reason as we have that in pinmap. We still need to update the time histogram when using heatmap. Should they share the same query? I think when we optimize pinmap and heatmap queries, their time histogram queries may also differ from each other, so I still keep them separate.

File: examples/twittermap/web/public/javascripts/map/controllers.js (hunk 6 above)
Comment (DeyuChen) [reply examples/twittermap/web/public/javascripts/map/controllers.js]: I changed it to randomizationSeed.

File: examples/twittermap/web/public/javascripts/map/controllers.js (hunk 7 above)
Comment (DeyuChen) [reply examples/twittermap/web/public/javascripts/map/controllers.js]: Done

File: examples/twittermap/web/public/javascripts/map/heatmap.js (hunk 8 above)
Comment (DeyuChen) [reply examples/twittermap/web/public/javascripts/map/heatmap.js]: Done

File: examples/twittermap/web/public/javascripts/map/heatmap.js (hunk 9 above)
Comment (DeyuChen) [reply examples/twittermap/web/public/javascripts/map/heatmap.js]: Done

**Comment by DeyuChen** [comment]
@baiqiushi I have already tried to avoid as much warnings in the Codacy/PR Quality Review as possible. I didn't change the single quotes to double quotes for consistency issue. And many warnings is related to the heatmap library I added in this PR, which I do not plan to modify at all.

File: examples/twittermap/web/public/javascripts/common/services.js (hunk 3 above)
Comment (waans11) [reply examples/twittermap/web/public/javascripts/common/services.js]: OK

File: examples/twittermap/web/public/javascripts/common/services.js (hunk 5 above)
Comment (waans11) [reply examples/twittermap/web/public/javascripts/common/services.js]: OK

File: examples/twittermap/web/public/javascripts/map/heatmap.js [hunk 10]
```diff
@@ -1,28 +1,119 @@
...
+        $scope.map.addLayer($scope.heatMapLayer);
+      }
+        
+      var points = [];
+      for (var i = 0; i < result.length; i++) {
+        if (result[i].hasOwnProperty('coordinate')){
+          points.push([result[i].coordinate[1], result[i].coordinate[0]]);
+        }
+        else {
+          points.push([$scope.rangeRandom(result[i].id, result[i]["place.bounding_box"][0][1], result[i]["place.bounding_box"][1][1]), $scope.rangeRandom(result[i].id + 79, result[i]["place.bounding_box"][0][0], result[i]["place.bounding_box"][1][0])]);
```
Comment (waans11) [inline examples/twittermap/web/public/javascripts/map/heatmap.js]: Why adding 79?

File: examples/twittermap/web/public/javascripts/map/pointmap.js [hunk 11]
```diff
@@ -291,7 +272,7 @@ angular.module('cloudberry.map')
             $scope.points.push([result[i].coordinate[1], result[i].coordinate[0]]);
           }
           else if (result[i].hasOwnProperty('place.bounding_box')){
-            $scope.points.push([rangeRandom(result[i].id, result[i]["place.bounding_box"][0][1], result[i]["place.bounding_box"][1][1]), rangeRandom(result[i].id + 79, result[i]["place.bounding_box"][0][0], result[i]["place.bounding_box"][1][0])]);
+            $scope.points.push([$scope.rangeRandom(result[i].id, result[i]["place.bounding_box"][0][1], result[i]["place.bounding_box"][1][1]), $scope.rangeRandom(result[i].id + 79, result[i]["place.bounding_box"][0][0], result[i]["place.bounding_box"][1][0])]);
```
Comment (waans11) [inline examples/twittermap/web/public/javascripts/map/pointmap.js]: Why adding 79?

File: examples/twittermap/web/public/javascripts/map/heatmap.js (hunk 10 above)
Comment (DeyuChen) [reply examples/twittermap/web/public/javascripts/map/heatmap.js]: It is just a magic number. We use tweet id as a random seed to generate longitude and latitude separately. If we use the same tweet id for the two values, they will always be the same and the resulting points will all be on the same orthogonal line. I will add a comment about it.

**Pull request merged**